### PR TITLE
chore: consistently run unit tests across beats

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
~~This PR uses the `mage unitTest` target across all Beats, adding them to all.~~

This PR adds a `unitTest` command in the Beats root project, which will reuse the existing `mage` build commands to run the unit tests for all Beats.

>~~I was tempted to modify the `libbeat/scripts/Makefile:L203` Make goal instead, to be used by all of them. But not sure of the implications of changing it there.~~

I also updated the docs to remove outdated info about Go's vendor directory, and the usage of certain build commands, basically replacing `make` with `mage`.

## Why is it important?
I ran `make unit` from the root dir of Beats and it failed because metricbeat does not have a `unit` make goal. Taking a look carefully to metricbeat's Makefile, it imports a different file than the other beats.

Apart from that, because other Beats used a `make unit` goal imported from libbeat, they did not run the `python unit tests`, failing the execution.

Because we want to make a consistent experience through the build system, we are adding this new command in `mage` instead of in `make`.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ] Verify how the unit tests are run on CI

## How to test this PR locally

```shell
cd beats
mage unitTest

# in a Beat
cd beats/$BEAT
mage unitTest
```

### CODE REVIEW

**Review by v1v — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

File: docs/devguide/contributing.asciidoc
Comment (jsoriano) [inline docs/devguide/contributing.asciidoc:67]:
```suggestion
eval $(gvm {go-version})
```

File: auditbeat/Makefile
Comment (jsoriano) [inline auditbeat/Makefile:17]: Do we want/need the make targets?

I guess the only reason to have them is to keep `make unit` working from the root directory, wdyt about replacing the call to `make unit` there with a call to the `mage` target for each project? 

If we want to have the same command in the root directory and in each project, instead of adding make targets to all projects, we could add a `mage unitTest` target to the root `magefile.go`, that calls this target on the individual projects.

**Review by jsoriano — COMMENTED** [review]
Looks good, but I think we should rather go in the direction of removing make targets.

File: auditbeat/Makefile
Comment (mdelapenya) [reply auditbeat/Makefile:17]: Agreed. Let's look for consistency in the _official_ build system

Comment (mdelapenya) [reply auditbeat/Makefile]: @jsoriano I updated the PR with the new `mage` goal, reverting the `make` goals. Please let me now what do you think, thanks!

File: magefile.go
Comment (jsoriano) [inline magefile.go:156]: This target will run all tests it finds with `go test`/`pytest` in the repository, this can be different to what the `make unit` test does, that loops over the list of projects.
If it works well I am on with the new approach, but just to be sure that we are aware of the possible differences.

File: auditbeat/Makefile
Comment (jsoriano) [reply auditbeat/Makefile]: Thanks!

File: magefile.go
Comment (mdelapenya) [reply magefile.go:156]: We could keep the `make unit` command as it is, reverting 54cd444d999570e4df070a57df715986d55d9e19, and start building a clear expectation from Mage side. I could think of a `test` namespace with:

- `test:unit`
- `test:integration`
- `test:system` 

Then we could add more child namespaces per beat 🤔

Comment (jsoriano) [reply magefile.go:156]: > reverting 54cd444

Is this target needed for this change? I think 54cd444 is fine.

But if this target is intended to replace `make unit`, it should probably do the same, looping over the projects list and run `mage unitTest` for each one of them. The result of running `go test` and `pytest` over all the repository may have the same result, but not sure.

If it is not intended to replace `make unit`, and is not needed for other targets, I think we could remove it by now.

> I could think of a `test` namespace with:
> 
>   * `test:unit`
>   * `test:integration`
>   * `test:system`
> 
> Then we could add more child namespaces per beat thinking

Yep, there was a proposal about doing this kind of namespacing, but we never renamed all the targets. I would leave this for a future change by now.

File: Makefile
Comment (jsoriano) [inline Makefile:61]: Btw, instead of changing directory I think you can use the `-d` flag.
```suggestion
	@$(foreach var,$(PROJECTS), mage -d $(CURDIR)/$(var) unitTest || exit 1;)
```

File: magefile.go
Comment (cachedout) [inline magefile.go:36]: Nit: This looks like it has an extra level of indentation. Is that intentional?

**Review by cachedout — APPROVED** [review]

File: magefile.go
Comment (jsoriano) [inline magefile.go:35]: `mage:import` is going to import all the targets defined in `unittest`, so there is no need to re-define `UnitTest` target if it only calls to `unittest.UnitTest`.

File: magefile.go
Comment (jsoriano) [reply magefile.go:36]: This is probably causing the linting failures.

**Comment by mdelapenya** [comment]
> @mdelapenya what should we with this PR? Shall we close it or update it?

Thanks for the ping. I think this PR is outdated and would need to be revisited with current state of the project. @jsoriano should we refresh this or close it?

**Comment by jsoriano** [comment]
> @jsoriano should we refresh this or close it?

I would leave the decision to @elastic/elastic-agent-data-plane, that are now the main users and maintainers of these targets. We should probably reassess if what was going to be solved by this PR is still needed.

I had concerns about the implementation of the `mage unitTest` target at the root level (see https://github.com/elastic/beats/pull/24787#discussion_r607968053). The team should decide if this is needed and how this should behave.

This PR also moved unit tests execution from make to mage, that I think is something nice to do.

If this is wanted I would propose to split this in two discussions/prs, one to discuss about the mage target at the root level, and another one to migrate the targets from make to mage.

**Comment by cmacknz** [comment]
I am in favour of migrating from `make` to `mage` here, we should have one way to do things.

We should be able to test for any differences between the proposed top level unit test target and the previous make target easily enough. 

I'm not sure how common it is to manually run the unit tests for every beat instead of letting CI do this validation.

**Comment by ruflin** [comment]
@cmacknz I wonder if your team should take over this PR from @mdelapenya ?

**Comment by cmacknz** [comment]
Yes, I had the same thought. I assigned it to myself to finish this.

**Comment by cmacknz** [comment]
I kept the documentation changes in https://github.com/elastic/beats/pull/31226

I decided not to port the mage unittest target, although I like the idea. Mostly I don't have time to verify that the target is equivalent to the make target and deal with any problems CI might find.

**Pull request closed without merging**
